framework/src/base/MooseApp.C

```diff
@@ -1340,24 +1347,18 @@ MooseApp::createMinimalApp()
 }
 
 void
-MooseApp::registerExecFlags()
+MooseApp::addExecFlag(const ExecFlagType & flag)
 {
-  registerExecFlag(EXEC_NONE);
-  registerExecFlag(EXEC_INITIAL);
-  registerExecFlag(EXEC_LINEAR);
-  registerExecFlag(EXEC_NONLINEAR);
-  registerExecFlag(EXEC_TIMESTEP_END);
-  registerExecFlag(EXEC_TIMESTEP_BEGIN);
-  registerExecFlag(EXEC_FINAL);
-  registerExecFlag(EXEC_FORCED);
-  registerExecFlag(EXEC_FAILED);
-  registerExecFlag(EXEC_CUSTOM);
-  registerExecFlag(EXEC_SUBDOMAIN);
-  registerExecFlag(EXEC_SAME_AS_MULTIAPP);
-}
-
-void
-MooseApp::registerExecFlag(const ExecFlagType & flag)
-{
-  Moose::execute_flags.addAvailableFlags(flag);
+  if (flag.id() == MooseEnumItem::INVALID_ID)
+  {
+    // It is desired that users when creating ExecFlagTypes should not worry about needing
+    // to assign a name and an ID. However, the ExecFlagTypes created by users are global constants
+    // and the ID to be assigned can't be known at construction time of this global constant, it is
+    // only known when it is added to this object (ExecFlagEnum). Therefore, this const cast allows
+    // the ID to be set after construction. This was the lesser of two evils: const_cast or
+    // friend class with mutable members.
+    ExecFlagType & non_const_flag = const_cast<ExecFlagType &>(flag);
```

Comment:
I can't quite understand why the ID can't be set when the name is... can you clarify?

Reply:
The ID's can be set, but this allows the ID to be optional. When we define flags it looks like:
```
const ExecFlagType EXEC_FOO("FOO");
```

@permcody didn't want to require that we also set the ids. Since, this is a const we need to do this cast to set the ID after it has been creted.

Reply:
I aree with that - I mean... why can't the ID be autoselected when the string is added.  Like use `size()` or something...